### ISSUE TITLE
Server RequestTimeout not working

### ISSUE THREAD

**Issue by amachadogarcia** [fastify/fastify#4008]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.29.0

### Node.js version

16.10.0

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

18.04.6

### Description

I'm trying to use the parameter requestTimeout when I declare the fastify instance. The issue that I found its fastify doesn't finish the request when the max time has been done. However, I didn't find anything related with that.

### Steps to Reproduce

    const app: FastifyInstance = fastify({});

    app.server.requestTimeout = 5000;

    app.register(RootRouteV1);

**Comment by mcollina** [comment]
The check is performed every 30 seconds, see https://nodejs.org/api/http.html#httpcreateserveroptions-requestlistener.

**Comment by mcollina** [comment]
cc @ShogunPanda

**Comment by ShogunPanda** [comment]
Actually it's reported on Node 16, so no `connectionsCheckingInterval` is in place.

@amachadogarcia do you have a way to reproduce it consistently? Is you route doing anything special?

**Comment by amachadogarcia** [comment]
Not really, my route has 5 hooks of fastify, onRequest, preParsing, preHandler, handler and preSerialization, and it works like a proxy, receives a request, processes it and through the handler send a request to another service. To reproduce it, I set the requestTimeout to 5 seconds when I initialize the fastify instance as I put in the in issue description, and then I put a promise to wait this 5 seconds on onRequest, as follows:
`await new Promise(f => setTimeout(f, 5000));`

**Comment by amachadogarcia** [comment]
Update: If I set the await promise in the onRequest I get a status code 400, I understand that the cause is because we don't have the body yet, but if for example I set the await promise on preHandler, the request timeout never works.

**Comment by mcollina** [comment]
Can you provide steps to reproduce? We often need a [reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), e.g. some code that allows someone else to recreate your problem by just copying and pasting it. If it involves more than a couple of different file, create a new repository on GitHub and add a link to that.

**Comment by amachadogarcia** [comment]
You have an example in this repository to reproduce it [reproducible example](https://github.com/amachadogarcia/Typescript-fastify-example). You only need compile it, start the service and make a request with any body and you will see that the option requestTimeout is not working.

**Comment by climba03003** [comment]
I think our behavior is the same as `node` core.
```js
const http = require("http");
const { setTimeout } = require("timers/promises");

const server = http.createServer(async function (request, reply) {
  console.time("request");
  await setTimeout(10000);
  console.timeEnd("request");
  reply.end("done");
});
server.requestTimeout = 5000;
server.listen(8082);
```

You can see that it will log `request: 10.001s` or something similar.

**Comment by climba03003** [comment]
AFAIK, `requestTimeout` require your request to be large enough that it transfer the payload more than `30s (default)`.
When the request meet the `onRequest` hook, it means the request transfer is finished. And the behavior should be expected.
@mcollina I think my understanding of `requestTimeout` is not wrong.

**Comment by ShogunPanda** [comment]
@climba03003 No, it is correct as you said. 

@amachadogarcia I'm afraid there is a misunderstanding of what `requestTimeout` is about. It is not a timeout on how much it takes for a request to be processed by fastify, but how much it takes for the underlying HTTP server to receive the request from the body. After it has been received and forwarded to fastify, there's no constraint (at least enforced from `requesTimeout`) on how much it takes for fastify hooks to process that data.

**Comment by amachadogarcia** [comment]
Yes, you are correct, I misunderstanded about how work requestTimeout so far, sorry, I was trying to close the connection in the middle of a hook, because eventually we get high wait times and I though that fastify had a way to do that. 

Thanks again for your time.

**Comment by ShogunPanda** [comment]
Unfortunately fastify has no way. What you can do is something like this:

```
preHandler(request, reply, done) => {
   const completed = false

   const timeout = setTimeout(() => {
      if(completed) {
        return
      }

      completed = true
      reply.code(500).send(/* ... */)
   }, 5000);

   doSomethingExpensive()
      .then(() => {
         if(completed) {
           return
         }
 
         completed = true
         done()
      })
})
```

Note that with this approach the expensive function will not be stopped but the client will receive a response within 5 seconds.

**Comment by climba03003** [comment]
I think someone created a plugin for this usage. It terminate the request if it wait too long.
https://github.com/metcoder95/fastify-racing

**Comment by amachadogarcia** [comment]
Ok, perfect, thanks for the info, I'll check it out.
